specs/advanced-data-layouts/prollytree/spec.md

```diff
@@ -0,0 +1,467 @@
+
+The `cursor` should be pointing to the node in a tree that rebalancing should start at. The root of the `cursor` will be upated to match the changes.
+
+The returned `ProllyNode` is for the new root of the tree.
+
+TODO: Account for `cursor.parent` being null
+TODO: Advance a cursor instead of iterating through `cursor.node.keys`
+TODO: Account for `cursor.node.keys` being empty (remove from parent)
+
+- Loop
+  - If `ShouldCreateBoundry(cursor.node.chunkconfig, cursor.node, CursorGetKey(cursor))` is `false`
+	- Call `AdvanceCursor(cursor)`
+	- continue to next loop cycle
+  - Create a new `ProllyNode` `node` by cloning the current `cursor.node`
+  - Remove all keys, values, links from `cursor.node` after `cursor.index`
+  - Remove all keys, values, links from `node` before and including `cursor.index`
+  - Generate a new `CID` for `cursor.node`
+  - Set the `cursor.parent.links` at `cursor.parent.index` to the new `CID`
+  - Generate a new `CID` for the new `node`
+  - Shift all items in `cursor.parent.keys` up by `1`
+  - Shift all items in `cursor.parent.links` up by `1`
+  - Increment `cursor.parent.index` by `1`
+  - Set the `cursor.parent.links` at `cursor.parent.index` to the new `CID`
+  - Set the `cursor.parent.keys` at `cursor.parent.index` to node.keys[0]
+  - Set cursor.node to `node`
+  - Set cursor.index to `0`
+- Return `RebalanceTree(cursor.parent)`
+
+### Put(ProllyNode, key, value) : ProllyNode
+
+- Get `cursor` from `CursorAtItem(ProllyNode, key)`
+- If `CursorIsValid(cursor)` is `true`
+	- check if the key is at `CursorGetKey`
+	- If it is
+		- Set the value in the `cursor.node.values` at `cursor.index` to `value`
+	- If it isn't
+		- add the `key` to `cursor.node.keys` after the `cursor.index`, shifting subsequent items down
+		- add the `values` to `cursor.node.values` after the `cursor.index`, shifting subsequent items down
+- TODO: If false, how is this reached? It means that there were no keys even "close" to `key`?
+- return `RebalanceTree(cursor)`
+
+### Delete(ProllyNode, key, value) : ProllyNode
+
+- Get `cursor` from `CursorAtItem(ProllyNode, key)`
+- If `CursorIsValid(cursor)` is `false`
+	- Return the `ProllyNode`
+- check if the key is at `CursorGetKey(cursor)`
+- If it is
+	- Remove the key in `cursor.node.keys` at `cursor.index`
+	- Remove the value in `cursor.node.values` at `cursor.index`
+- return `RebalanceTree(cursor)`
+
+### Search(ProllyNode, prefix) : Iterator<key, value>
+
+- Get `cursor` from `CursorAtItem(ProllyNode, key)`
+- Create an Iterator (language dependent)
+- On each pull from the iterator
+	- If `CursorIsValid(cursor)` is `false`
+		- Close the iterator and return
+	- Get the `key` from `CursorGetKey(cursor)`
+	- If `key` does not start with `prefix`
+		- Close the iterator and return
+	- Get the `value` from `CursorGetValue(cursor)`
+	- Yield the `key` and `value` from the iterator
+	- AdvanceCursor(ProllyNode)
+
+### Diff(ProllyNode base, ProllyNode new) : Iterator< Diff >
```

Comment:
Should the function be public api or just used in the merge function?

Reply:
I don't think we need to perscribe whether something is public or not in the spec. But IMO it should be a public API in implementations since it's generally useful to have.